cpu/atmega_common/periph/gpio.c

```diff
@@ -254,6 +428,77 @@ static inline void irq_handler(uint8_t int_num)
     __exit_isr();
 }
 
+#ifdef PCINT_NUM_BANKS
+/* inline function that is used by the PCINT ISR */
+static inline void pcint_handler(uint8_t bank, volatile uint8_t *mask_reg)
+{
+    __enter_isr();
+    /* Find right item */
+    uint8_t pin_num = 0;
+    uint8_t enabled_pcints = *mask_reg;
+
+    while (enabled_pcints > 0) {
+        /* check if this pin is enabled & has changed */
+        if (enabled_pcints & 0x1) {
+            /* re-construct mask from pin number*/
+            uint8_t pin_mask = (1 << pin_num);
+            /* get pin from mapping (assumes 8 entries per bank!) */
+            gpio_t pin = pcint_mapping[ bank * 8 + pin_num ];
+            uint8_t port_value = (_SFR_MEM8(_pin_addr( pin )));
+            uint8_t pin_value = ((port_value & pin_mask) != 0);
+            uint8_t old_state = ((pcint_state[ bank ] & pin_mask) != 0);
+            gpio_isr_ctx_pcint_t *conf = &pcint_config[ bank * 8 + pin_num ];
+            if (old_state != pin_value) {
+                if (pin_value) {
+                    pcint_state[ bank ] |= (pin_mask);
+                }
+                else {
+                    pcint_state[ bank ] &= ~(pin_mask);
+                }
```

Comment:
```C
                if (pin_value) {
                    pcint_state[ bank ] |= (pin_mask);
                }
                else {
                    pcint_state[ bank ] &= ~(pin_mask);
                }
```
This could be simplified to `pcint_state[bank] ^= pin_mask;`, which will just toggle the bit. That would be correct for both cases

Reply:
The idea is *not* to toggle the bit, but to set the bit correctly to 1 or 0. Using your syntax does not work here!

Reply:
This is only entered if the current value of the pin is different from the value read from the pin differs from the value stored in `pcint_state`.

If `pin_value` is not 0 (then-case of the inner `if`) and also `pin_value` differs from what it is stored (then-case of outer `if`), then the pin is set. If both `then`-branches are taken, the value in `pcint_state` for that bit has to be zero and is not set. This is equivalent to toggling.

if `pin_value` is zero (else-case of inner `if`) and differs from what is stored in `pcint_state` (then-case of outer if), than the bit is clear. But to take the then-branch of the outer `if` and the else-branch of the inner `if`, the corresponding bit had to be `1`. So this is again equivalent to toggling the bit.

In both cases the bit is just toggled, so the code could be written like that to be more efficient. (If you don't believe me: Change it to toggling and run the test ;-))

Reply:
I fixed to typos above that made understanding what I meant impossible. It should now make sense.

Reply:
In the latter case for the `!=` I agree, then we can toggle. For the initial state you can't!